### ISSUE TITLE
SavedModel is not deterministic when saved with enable_op_determinism=True

### ISSUE THREAD

**Issue by Krasner** [tensorflow/tensorflow#61131]
### Issue type

Support

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.9.2

### Custom code

No

### OS platform and distribution

Ubuntu 20.04

### Current behavior?

I have a model with monte-carlo sampling layers. At train time it is ok for sampling to be random and I do not set a seed.
At inference this model MUST output deterministic results - so the sampled values must be reproducible.
I need to save this model after training using `tf.keras.models.save_model()` and then load it with `tf.keras.models.load_model()`

When training is complete I save the model like so:
```
tf.keras.utils.set_random_seed(42)
tf.config.experimental.enable_op_determinism()
model.seed = 42
tf.keras.models.save_model(model, save_path, save_format="tf")
```

To load it:
```
import tensorflow as tf
tf.keras.utils.set_random_seed(42)
tf.config.experimental.enable_op_determinism()
model = tf.keras.models.load_model(save_path)
for _ in range(10):
 output = model(input_tensor, training=False)
```
I expect output to be always the same for the same input - within one python process and when reloading, however it is not.

This is what my model looks like (only the relevant portions):

```
class CustomModel(tf.keras.Model):
   def __init__(self, *args, **kwargs):
       self._seed = None
       self.mc_layer = CustomMonteCarloSamplingLayer()

    @tf.keras.utils.register_keras_serializable(name="seed")
    @property
    def seed(self):
        return self._seed
     
    @seed.setter
    def seed(self, value):
        if is_op_determinism_enabled():
            lgr.info("Setting random seed")
            self._seed = value
            tf.random.set_seed(value)
        else:
            lgr.info("TF OP Determinism not set")

   def call(self, inputs, training=True):
       return self.mc_layer(inputs, seed=self._seed)
```

Internally in `CustomMonteCarloSamplingLayer` I use the seed value like so:
```
tf.random.set_seed(seed)
samples= dist.sample(num_samples, seed=seed)
```
where dist is a distribution from tensorflow_probability.

When `self._seed = None` which is the case in training, `dist.sample` will randomly generate values.
When `self._seed` is set then I need `dist.sample` to be reproducible.

The loaded model, even with op determinism enabled during saving and loading, is not reproducible.

      

### Standalone code to reproduce the issue

```shell
See above
```

**Comment by Krasner** [comment]
To clarify, in this saving block:
```
tf.keras.utils.set_random_seed(42)
tf.config.experimental.enable_op_determinism()
model.seed = 42

# Model is deterministic here

tf.keras.models.save_model(model, save_path, save_format="tf")
```

So, before saving the model it is acting deterministically after the seed update, but after the save (and upon reloading) the determinism is lost

**Comment by Krasner** [comment]
I have an update which might be a solution.
If I set: `self._seed = tf.Variable([42, 42], trainable=False)`

Then tfp.distribution will sample reproducibly
```
# tf.random.set_seed(seed)
samples= dist.sample(num_samples, seed=seed) 
```
Note: I don't use `tf.random.set_seed` instead passing the size 2 seed to `dist.sample` because of the underlying `tf.random.stateless_xxxx` function

**Comment by sushreebarsa** [comment]
@Krasner Sorry for the late response!
Could you please share the code in colab gist or notebook as it seems that some dependencies are missing here. Also please refer to this [link](https://www.tensorflow.org/api_docs/python/tf/keras/saving/save_model) to save the model and let us know if it helps?
Thank you!
